File: content/examples/other/leetcode/problems/graph-valid-tree/example-2/Solution.java
public class Solution {
  static class UnionFind {
    int[] parent;
    int[] rank;
    int components;

    UnionFind(int n) {
      parent = new int[n];
      rank = new int[n];
      components = n;
      for (int i = 0; i < n; i += 1) parent[i] = i;
    }

    int find(int x) {
      while (parent[x] != x) {
        parent[x] = parent[parent[x]];
        x = parent[x];
      }
      return x;
    }

    boolean union(int a, int b) {
      int ra = find(a);
      int rb = find(b);
      if (ra == rb) return false;
      if (rank[ra] < rank[rb]) {
        int tmp = ra;
        ra = rb;
        rb = tmp;
      }
      parent[rb] = ra;
      if (rank[ra] == rank[rb]) rank[ra] += 1;
      components -= 1;
      return true;
    }
  }

  static boolean validTreeUnionFind(int n, int[][] edges) {
    UnionFind uf = new UnionFind(n);
    for (int[] e : edges) {
      if (!uf.union(e[0], e[1])) return false;
    }
    return uf.components == 1;
  }

  public static void main(String[] args) {
    System.out.println(validTreeUnionFind(5, new int[][] {{0, 1}, {0, 2}, {0, 3}, {1, 4}}));
    System.out.println(
        validTreeUnionFind(5, new int[][] {{0, 1}, {1, 2}, {2, 3}, {1, 3}, {1, 4}}));
  }
}

File: content/examples/other/leetcode/problems/graph-valid-tree/example-2/solution.js
class UnionFind {
  constructor(n) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.rank = new Array(n).fill(0);
    this.components = n;
  }

  find(x) {
    let cur = x;
    while (this.parent[cur] !== cur) {
      this.parent[cur] = this.parent[this.parent[cur]];
      cur = this.parent[cur];
    }
    return cur;
  }

  union(a, b) {
    let ra = this.find(a);
    let rb = this.find(b);
    if (ra === rb) return false;
    if (this.rank[ra] < this.rank[rb]) [ra, rb] = [rb, ra];
    this.parent[rb] = ra;
    if (this.rank[ra] === this.rank[rb]) this.rank[ra] += 1;
    this.components -= 1;
    return true;
  }
}

function validTreeUnionFind(n, edges) {
  const uf = new UnionFind(n);
  for (const [u, v] of edges) {
    if (!uf.union(u, v)) return false;
  }
  return uf.components === 1;
}

if (require.main === module) {
  console.log(validTreeUnionFind(5, [[0, 1], [0, 2], [0, 3], [1, 4]]));
  console.log(validTreeUnionFind(5, [[0, 1], [1, 2], [2, 3], [1, 3], [1, 4]]));
}

module.exports = { validTreeUnionFind };
